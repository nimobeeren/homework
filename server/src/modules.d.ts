/*
 * Here we declare any modules that do not have type definitions available.
 */

type DocumentNode = import('graphql').DocumentNode;

/**
 * GraphQL schema file (as loaded by Parcel).
 */
declare module '*.graphql' {
  const content: DocumentNode;
  export default content;
}
