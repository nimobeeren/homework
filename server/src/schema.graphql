type Video {
  title: String
  path: String
}
