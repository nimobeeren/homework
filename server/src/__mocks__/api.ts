export const resolvers = {
  Query: {
    foo: () => 'bar'
  }
};
