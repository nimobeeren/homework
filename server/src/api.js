const resolvers = {
  Query: {
    videos: () => [{ title: 'Hello World', path: '/hello/world.mp4' }]
  }
};

module.exports = resolvers;
