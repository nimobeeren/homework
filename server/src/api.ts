export const resolvers = {
  Query: {
    files: () => [
      { title: 'Hello World', path: '/hello/world.mp4' },
      { title: 'Breaking Bad', path: '/hello/breaking-bad.mp4' }
    ]
  }
};
