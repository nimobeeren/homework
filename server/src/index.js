const path = require('path');
const fs = require('fs');
const Koa = require('koa');
const koaWebpack = require('koa-webpack');
const serve = require('koa-static');
const { ApolloServer } = require('apollo-server-koa');
const webpackDevConfig = require('../../webpack.dev');
const webpackProdConfig = require('../../webpack.prod');
const resolvers = require('./api');

const PORT = process.env.PORT || 3000;

async function start() {
  const app = new Koa();

  // Register API middleware
  const typeDefs = fs.readFileSync(path.resolve(__dirname, 'schema.graphql'), 'utf8');
  const server = new ApolloServer({ typeDefs, resolvers });
  server.applyMiddleware({ app });

  // Serve client
  if (process.env.NODE_ENV !== 'production') {
    // Use middleware to support HMR.
    const devMiddleware = await koaWebpack({ config: webpackDevConfig });
    app.use(devMiddleware);

    // Serve HTML from memory, generated by HTML Webpack Plugin
    app.use(async ctx => {
      const filename = path.resolve(webpackDevConfig.output.path, 'index.html');
      const devFs = devMiddleware.devMiddleware.fileSystem;
      ctx.response.type = 'html';
      ctx.response.body = devFs.createReadStream(filename);
    });
  } else {
    // Serve static build
    console.log('Running in production mode');
    const staticPath = webpackProdConfig.output.path;
    app.use(serve(path.resolve(staticPath)));
  }

  app.listen(PORT);
  console.info(`ðŸš€ Server listening on port ${PORT}`);
}

start();
