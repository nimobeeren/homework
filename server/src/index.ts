import path from 'path';
import fs from 'fs';
import Koa from 'koa';
import koaWebpack from 'koa-webpack';
import serve from 'koa-static';
import { ApolloServer } from 'apollo-server-koa';
import { webpackDevConfig } from '../../webpack.dev';
import { webpackProdConfig } from '../../webpack.prod';
import { resolvers } from './api';

const PORT = process.env.PORT || 3000;

async function start(): Promise<void> {
  const app = new Koa();

  // Register API middleware
  const typeDefs = fs.readFileSync(path.resolve(__dirname, 'schema.graphql'), 'utf8');
  const server = new ApolloServer({ typeDefs, resolvers });
  server.applyMiddleware({ app });

  // Serve client
  if (process.env.NODE_ENV !== 'production') {
    // Use middleware to support HMR
    const devMiddleware = await koaWebpack({ config: webpackDevConfig });
    app.use(devMiddleware);

    // Serve HTML from memory, generated by HTML Webpack Plugin
    app.use(async ctx => {
      const filename = path.resolve(webpackDevConfig.output.path, 'index.html');
      const devFs = devMiddleware.devMiddleware.fileSystem;
      ctx.response.type = 'html';
      ctx.response.body = devFs.createReadStream(filename);
    });
  } else {
    // Serve static build
    console.log('Running in production mode');
    const staticPath = webpackProdConfig.output.path;
    app.use(serve(path.resolve(staticPath)));
  }

  app.listen(PORT);
  console.info(`ðŸš€ Server listening on port ${PORT}`);
}

start();
