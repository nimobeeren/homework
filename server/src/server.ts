import path from 'path';
import Koa from 'koa';
import serve from 'koa-static';
import { ApolloServer } from 'apollo-server-koa';
import cors from '@koa/cors';
import typeDefs from '../../graphql/schema.graphql';
import { resolvers } from './api';

export function start(port: number | string) {
  const app = new Koa();

  // Register API middleware
  const apollo = new ApolloServer({ typeDefs, resolvers });
  apollo.applyMiddleware({ app });

  // Enable CORS to allow access from client
  app.use(cors());

  // Serve static client build in production
  if (process.env.NODE_ENV === 'production') {
    console.log('Running in production mode');
    const staticPath = path.resolve(__dirname, '../../client/dist');
    app.use(serve(staticPath));
  }

  const server = app.listen(port);
  console.info(`ðŸš€  Server listening on port ${port}`);
  return server;
}
