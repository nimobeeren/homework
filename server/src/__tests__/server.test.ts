import { Server } from 'http';
import { ApolloServer } from 'apollo-server-koa';
import { createTestClient } from 'apollo-server-testing';
import request from 'supertest';
import typeDefs from '../../../graphql/src/schema.graphql';
import { resolvers } from '../api';
import { FileListDocument } from '../../../graphql/src';
import { start } from '../server';

jest.mock('../../../graphql/src/schema.graphql');
jest.mock('../api');

describe('Server', () => {
  let httpServer: Server;
  beforeAll(() => {
    httpServer = start(process.env.PORT || 4000);
  });
  afterAll(() => {
    httpServer.close();
  });

  it('responds to HTTP requests on the GraphQL endpoint', async () => {
    const response = await request(httpServer).post('/graphql');
    // We don't care exactly what the response is, as long as we get one
    // The implementation of the GraphQL server is not our responsiblity
    expect(response.status).toMatchSnapshot();
  });

  it('serves a file list', async () => {
    // create a test server to test against
    const apolloServer = new ApolloServer({ typeDefs, resolvers });

    // use the test server to create a query function
    const { query } = createTestClient(apolloServer);

    // run query against the server and snapshot the output
    const res = await query({ query: FileListDocument });
    expect(res).toMatchSnapshot();
  });
});
