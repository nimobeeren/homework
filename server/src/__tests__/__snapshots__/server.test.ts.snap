// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Server responds to HTTP requests on the GraphQL endpoint 1`] = `500`;

exports[`Server serves a file list 1`] = `
Object {
  "data": Object {
    "files": Object {
      "items": Array [
        Object {
          "path": "/hello/world.mp4",
          "title": "Hello World",
        },
        Object {
          "path": "/hello/breaking-bad.mp4",
          "title": "Breaking Bad",
        },
      ],
    },
  },
  "errors": undefined,
  "extensions": undefined,
  "http": Object {
    "headers": Headers {
      Symbol(map): Object {},
    },
  },
}
`;
